File: tests/test_nodes/test_volatile.py
# Author: Nicolas Legrand <nicolas.legrand@cas.au.dk>

import numpy as np
import pytest
from pyhgf.rshgf import Network as RsNetwork

from pyhgf import load_data
from pyhgf.model import Network as PyNetwork

UPDATE_TYPES = ["standard", "eHGF", "unbounded"]


def _assert_value_level_match(net_a, node_a, net_b, node_b, label="", rtol=1e-5):
    """Assert value-level trajectories match between two networks."""
    for key in ["mean", "expected_mean", "precision", "expected_precision"]:
        assert np.allclose(
            net_a.node_trajectories[node_a][key],
            net_b.node_trajectories[node_b][key],
            rtol=rtol,
        ), f"{label}: Value-level key '{key}' mismatch"


def _assert_vol_level_match(
    volatile_net, vol_node, explicit_net, exp_node, label="", rtol=1e-4, atol=1e-6
):
    """Assert volatility-level trajectories match (volatile _vol vs explicit node)."""
    vol_key_map = {
        "mean_vol": "mean",
        "expected_mean_vol": "expected_mean",
        "precision_vol": "precision",
        "expected_precision_vol": "expected_precision",
    }
    for vol_key, explicit_key in vol_key_map.items():
        assert np.allclose(
            volatile_net.node_trajectories[vol_node][vol_key],
            explicit_net.node_trajectories[exp_node][explicit_key],
            rtol=rtol,
            atol=atol,
        ), f"{label}: Volatility-level key '{vol_key}' vs '{explicit_key}' mismatch"


def _build_volatile(cls, volatility_updates, timeseries, mean_field_updates=False):
    """Build a volatile-state network with autoconnection for equivalence testing."""
    return (
        cls(
            volatility_updates=volatility_updates, mean_field_updates=mean_field_updates
        )
        .add_nodes()
        .add_nodes(
            kind="volatile-state",
            value_children=0,
            autoconnection_strength=1.0,
        )
        .input_data(input_data=timeseries)
    )


def _build_explicit(cls, volatility_updates, timeseries, mean_field_updates=False):
    """Build explicit continuous + volatility-parent network."""
    return (
        cls(
            volatility_updates=volatility_updates, mean_field_updates=mean_field_updates
        )
        .add_nodes()
        .add_nodes(value_children=0)
        .add_nodes(volatility_children=1)
        .input_data(input_data=timeseries)
    )


def _run_volatile_vs_explicit(volatility_updates):
    """Test that volatile-state is equivalent to explicit continuous+vol-parent pair.

    Both the Python and Rust backends satisfy this equivalence: the value-level
    posterior update runs before the volatility-level prediction-error step in both.
    """
    timeseries = load_data("continuous")

    for cls, label in [
        (PyNetwork, f"{volatility_updates} py"),
        (RsNetwork, f"{volatility_updates} rs"),
    ]:
        vol = _build_volatile(cls, volatility_updates, timeseries)
        exp = _build_explicit(cls, volatility_updates, timeseries)

        _assert_value_level_match(vol, 0, exp, 0, f"{label} input")
        _assert_value_level_match(vol, 1, exp, 1, label)
        _assert_vol_level_match(vol, 1, exp, 2, label)


def _run_explicit_cross_backend(volatility_updates):
    """Test that Python and Rust produce the same trajectories for explicit networks."""
    timeseries = load_data("continuous")

    exp_py = _build_explicit(PyNetwork, volatility_updates, timeseries)
    exp_rs = _build_explicit(RsNetwork, volatility_updates, timeseries)

    label = f"{volatility_updates} py vs rs"
    # Unbounded path: see docstring — Python and Rust use mathematically
    # equivalent but float-distinct unbounded posterior kernels.
    rtol = 1e-1 if volatility_updates == "unbounded" else 1e-4
    _assert_value_level_match(exp_py, 0, exp_rs, 0, f"{label} input", rtol=rtol)
    _assert_value_level_match(exp_py, 1, exp_rs, 1, label, rtol=rtol)


def test_volatile_node_matches_explicit_volatility_parent():
    """Test Rust volatile-state against explicit continuous+vol-parent (standard)."""
    _run_volatile_vs_explicit("standard")


def test_volatile_node_ehgf_matches_explicit():
    """Test Rust volatile-state against explicit continuous+vol-parent (eHGF)."""
    _run_volatile_vs_explicit("eHGF")


def test_volatile_node_unbounded_matches_explicit():
    """Test Rust volatile-state against explicit continuous+vol-parent (unbounded)."""
    _run_volatile_vs_explicit("unbounded")


def test_explicit_cross_backend_standard():
    """Test Python and Rust explicit networks agree (standard update)."""
    _run_explicit_cross_backend("standard")


def test_explicit_cross_backend_ehgf():
    """Test Python and Rust explicit networks agree (eHGF update)."""
    _run_explicit_cross_backend("eHGF")


def test_explicit_cross_backend_unbounded():
    """Test Python and Rust explicit networks agree (unbounded update)."""
    _run_explicit_cross_backend("unbounded")


def _run_volatile_input_invariance(cls, label, mean_field_updates=False):
    """Vary ``tonic_volatility`` of a volatile-state input/leaf node.

    The input/leaf node has no value children, so it does not undergo a Gaussian random
    walk between observations. Its ``expected_precision`` must therefore be its prior
    precision — identical for any value of ``tonic_volatility`` — mirroring the
    continuous-node treatment.
    """
    timeseries = np.array([0.5, 1.0, 0.7, 0.3])
    expected_precisions = []
    for omega in [-8.0, 0.0]:
        net = (
            cls(
                volatility_updates="unbounded",
                mean_field_updates=mean_field_updates,
            )
            .add_nodes(
                kind="volatile-state",
                tonic_volatility=omega,
                precision=5.0,
                expected_precision=5.0,
            )
            .add_nodes(value_children=0)
            .input_data(input_data=timeseries)
        )
        expected_precisions.append(
            np.asarray(net.node_trajectories[0]["expected_precision"])
        )

    np.testing.assert_allclose(
        expected_precisions[0],
        expected_precisions[1],
        rtol=1e-5,
        err_msg=(
            f"{label}: input volatile-state node's expected_precision changes "
            "with tonic_volatility — input-leaf override missing"
        ),
    )
    # And it should equal the prior precision of 5.0 throughout.
    np.testing.assert_allclose(expected_precisions[0], 5.0, rtol=1e-5)


def test_volatile_input_node_invariant_to_tonic_volatility_python():
    """Per-node Python: volatile-state input ignores tonic_volatility."""
    _run_volatile_input_invariance(PyNetwork, "py")


def test_volatile_input_node_invariant_to_tonic_volatility_rust():
    """Per-node Rust: volatile-state input ignores tonic_volatility."""
    _run_volatile_input_invariance(RsNetwork, "rs")


def _assert_volatile_node_cross_backend(net_py, net_rs, node, label="", rtol=1e-4):
    """Assert a volatile node's trajectories match across backends."""
    keys = [
        "mean",
        "expected_mean",
        "precision",
        "expected_precision",
        "mean_vol",
        "expected_mean_vol",
        "precision_vol",
        "expected_precision_vol",
    ]
    for key in keys:
        assert np.allclose(
            net_py.node_trajectories[node][key],
            net_rs.node_trajectories[node][key],
            rtol=rtol,
        ), f"{label}: key '{key}' mismatch"


@pytest.mark.parametrize("volatility_updates", UPDATE_TYPES)
def test_volatile_mean_field_cross_backend(volatility_updates):
    """JAX and Rust agree for volatile-state nodes with ``mean_field_updates=True``.

    Exercises the mean-field prediction and posterior paths of the volatile-state node
    (the ``_mean_field`` update functions), which the relaxed-default tests above do not
    cover.
    """
    timeseries = load_data("continuous")

    vol_py = _build_volatile(
        PyNetwork, volatility_updates, timeseries, mean_field_updates=True
    )
    vol_rs = _build_volatile(
        RsNetwork, volatility_updates, timeseries, mean_field_updates=True
    )

    # Unbounded path: Python and Rust use mathematically equivalent but float-distinct
    # unbounded posterior kernels (see _run_explicit_cross_backend).
    rtol = 1e-1 if volatility_updates == "unbounded" else 1e-4
    _assert_volatile_node_cross_backend(
        vol_py, vol_rs, 1, f"{volatility_updates} mean_field py vs rs", rtol=rtol
    )


@pytest.mark.parametrize("volatility_updates", UPDATE_TYPES)
def test_explicit_mean_field_cross_backend(volatility_updates):
    """JAX and Rust agree for explicit continuous+vol-parent with mean-field updates."""
    timeseries = load_data("continuous")

    exp_py = _build_explicit(
        PyNetwork, volatility_updates, timeseries, mean_field_updates=True
    )
    exp_rs = _build_explicit(
        RsNetwork, volatility_updates, timeseries, mean_field_updates=True
    )

    label = f"{volatility_updates} mean_field py vs rs"
    rtol = 1e-1 if volatility_updates == "unbounded" else 1e-4
    _assert_value_level_match(exp_py, 0, exp_rs, 0, f"{label} input", rtol=rtol)
    _assert_value_level_match(exp_py, 1, exp_rs, 1, label, rtol=rtol)
    _assert_value_level_match(exp_py, 2, exp_rs, 2, label, rtol=rtol)


def test_volatile_input_node_invariant_mean_field_python():
    """Per-node Python: volatile-state input ignores tonic_volatility (mean-field)."""
    _run_volatile_input_invariance(PyNetwork, "py mean_field", mean_field_updates=True)


def test_volatile_input_node_invariant_mean_field_rust():
    """Per-node Rust: volatile-state input ignores tonic_volatility (mean-field)."""
    _run_volatile_input_invariance(RsNetwork, "rs mean_field", mean_field_updates=True)


def _build_volatile_value_parent(coupling_fn, mean_field_updates, timeseries):
    """Build a volatile-state value parent coupled to its child via ``coupling_fn``."""
    net = PyNetwork(
        volatility_updates="standard", mean_field_updates=mean_field_updates
    ).add_nodes()
    if coupling_fn is None:
        net = net.add_nodes(
            kind="volatile-state", value_children=0, autoconnection_strength=1.0
        )
    else:
        net = net.add_nodes(
            kind="volatile-state",
            value_children=0,
            autoconnection_strength=1.0,
            coupling_fn=coupling_fn,
        )
    return net.input_data(input_data=timeseries)


@pytest.mark.parametrize("mean_field_updates", [False, True])
def test_volatile_nonlinear_coupling(mean_field_updates):
    """Volatile-state node with non-linear (tanh) value coupling (JAX).

    Exercises the non-linear coupling branches (the ``grad`` / ``grad(grad(...))``
    terms) of the volatile-state value-level precision and mean updates, in both the
    relaxed and mean-field schemes. The trajectories must stay finite, and the
    non-linear coupling must actually change the result relative to linear coupling.

    The two backends are not compared here: the ``tanh``-coupled volatile node (with
    autoconnection and an implicit volatility level) is numerically sensitive, so the
    JAX and Rust trajectories diverge over time despite using the same update rule.
    """
    import jax.numpy as jnp

    timeseries = load_data("continuous")
    nonlinear = _build_volatile_value_parent(
        (jnp.tanh,), mean_field_updates, timeseries
    )
    linear = _build_volatile_value_parent(None, mean_field_updates, timeseries)

    label = f"nonlinear mean_field={mean_field_updates}"
    for key in ["mean", "expected_mean", "precision", "expected_precision"]:
        values = np.asarray(nonlinear.node_trajectories[1][key])
        assert np.all(np.isfinite(values)), f"{label}: non-finite '{key}'"

    # The tanh coupling must change the value-level trajectory relative to linear.
    assert not np.allclose(
        nonlinear.node_trajectories[1]["mean"],
        linear.node_trajectories[1]["mean"],
    ), f"{label}: tanh coupling did not change the trajectory"


def _build_volatile_value_child(coupling_fn, mean_field_updates, timeseries):
    """Build a volatile-state node that is the value *child* of a volatile parent.

    The parent → child edge carries ``coupling_fn``, so the child's prediction step
    reads the parent's expected mean through that coupling. A volatile parent is used
    (rather than a continuous one) because continuous parents read their children's
    ``observed`` flag, which volatile-state nodes do not expose.
    """
    net = PyNetwork(
        volatility_updates="standard", mean_field_updates=mean_field_updates
    ).add_nodes()
    net = net.add_nodes(kind="volatile-state", value_children=0)  # node 1: child
    # The parent (node 2) carries its mean forward (autoconnection_strength=1.0) so it
    # drifts away from 0; otherwise its expected mean stays at 0, where tanh(0) == 0
    # is indistinguishable from linear coupling.
    if coupling_fn is None:
        net = net.add_nodes(
            kind="volatile-state", value_children=1, autoconnection_strength=1.0
        )
    else:
        net = net.add_nodes(
            kind="volatile-state",
            value_children=1,
            autoconnection_strength=1.0,
            coupling_fn=coupling_fn,
        )
    return net.input_data(input_data=timeseries)


@pytest.mark.parametrize("mean_field_updates", [False, True])
def test_volatile_nonlinear_value_parent(mean_field_updates):
    """Volatile-state node with a non-linearly-coupled value parent (JAX).

    Exercises the prediction-step non-linear coupling branches
    (:func:`predict_mean_value_level` and the relaxed
    :func:`predict_precision_value_level`) for a volatile node that sits *below* a
    ``tanh``-coupled value parent. The trajectories must stay finite, and the non-linear
    coupling must change the result relative to linear coupling.
    """
    import jax.numpy as jnp

    timeseries = load_data("continuous")
    nonlinear = _build_volatile_value_child((jnp.tanh,), mean_field_updates, timeseries)
    linear = _build_volatile_value_child(None, mean_field_updates, timeseries)

    label = f"nonlinear value parent mean_field={mean_field_updates}"
    for key in ["mean", "expected_mean", "precision", "expected_precision"]:
        values = np.asarray(nonlinear.node_trajectories[1][key])
        assert np.all(np.isfinite(values)), f"{label}: non-finite '{key}'"

    # The tanh coupling on the parent edge must change the child's trajectory.
    assert not np.allclose(
        nonlinear.node_trajectories[1]["mean"],
        linear.node_trajectories[1]["mean"],
    ), f"{label}: tanh coupling did not change the trajectory"

File: tests/test_volatile.rs
use rshgf::model::Network;
use std::collections::HashMap;

/// Helper to assert approximate equality of f64 values.
fn assert_close(actual: f64, expected: f64, tol: f64, label: &str) {
    assert!(
        (actual - expected).abs() < tol,
        "{}: expected {}, got {} (diff = {})",
        label,
        expected,
        actual,
        (actual - expected).abs()
    );
}

/// Assert that the value-level trajectories of two nodes in two networks match.
fn assert_value_level_match(
    net_a: &Network,
    node_a: usize,
    net_b: &Network,
    node_b: usize,
    label: &str,
) {
    let traj_a = &net_a.node_trajectories.nodes[node_a];
    let traj_b = &net_b.node_trajectories.nodes[node_b];

    let fields: Vec<(&str, &Vec<f64>, &Vec<f64>)> = vec![
        ("mean", &traj_a.mean, &traj_b.mean),
        ("expected_mean", &traj_a.expected_mean, &traj_b.expected_mean),
        ("precision", &traj_a.precision, &traj_b.precision),
        ("expected_precision", &traj_a.expected_precision, &traj_b.expected_precision),
    ];

    for (key, a, b) in &fields {
        for (t, (va, vb)) in a.iter().zip(b.iter()).enumerate() {
            assert_close(
                *va,
                *vb,
                1e-6,
                &format!("{} value-level '{}' t={}", label, key, t),
            );
        }
    }
}

/// Assert that the volatility-level trajectories of a volatile node match the
/// corresponding explicit node trajectories.
fn assert_vol_level_match(
    volatile_net: &Network,
    vol_node: usize,
    explicit_net: &Network,
    exp_node: usize,
    label: &str,
) {
    let vol_traj = &volatile_net.node_trajectories.nodes[vol_node];
    let exp_traj = &explicit_net.node_trajectories.nodes[exp_node];

    let fields: Vec<(&str, &Vec<f64>, &str, &Vec<f64>)> = vec![
        ("mean_vol", &vol_traj.mean_vol, "mean", &exp_traj.mean),
        ("expected_mean_vol", &vol_traj.expected_mean_vol, "expected_mean", &exp_traj.expected_mean),
        ("precision_vol", &vol_traj.precision_vol, "precision", &exp_traj.precision),
        ("expected_precision_vol", &vol_traj.expected_precision_vol, "expected_precision", &exp_traj.expected_precision),
    ];

    for (vol_key, a, exp_key, b) in &fields {
        for (t, (va, vb)) in a.iter().zip(b.iter()).enumerate() {
            assert_close(
                *va,
                *vb,
                1e-6,
                &format!("{} vol-level '{}' vs '{}' t={}", label, vol_key, exp_key, t),
            );
        }
    }
}

/// Build a volatile network: input (node 0) + volatile-state value parent (node 1).
fn build_volatile_network(update_type: &str, data: &[f64]) -> Network {
    let mut net = Network::new(update_type);
    net.add_nodes("continuous-state", 1, None, None, None, None, None, None);
    net.add_nodes("volatile-state", 1, None, Some(0.into()), None, None, None,
        Some(HashMap::from([("autoconnection_strength".into(), 1.0)])));
    net.set_update_sequence();
    net.input_data(data.iter().map(|v| vec![*v]).collect(), None, true);
    net
}

/// Build an explicit network: input (node 0) + value parent (node 1) + volatility
/// parent of node 1 (node 2).
fn build_explicit_network(update_type: &str, data: &[f64]) -> Network {
    let mut net = Network::new(update_type);
    net.add_nodes("continuous-state", 1, None, None, None, None, None, None);
    net.add_nodes("continuous-state", 1, None, Some(0.into()), None, None, None, None);
    net.add_nodes("continuous-state", 1, None, None, None, Some(1.into()), None, None);
    net.set_update_sequence();
    net.input_data(data.iter().map(|v| vec![*v]).collect(), None, true);
    net
}

/// Run the volatile-vs-explicit comparison for the given update type.
fn compare_volatile_and_explicit(update_type: &str) {
    let data: Vec<f64> = (0..20).map(|i| (i as f64) * 0.1).collect();

    let volatile_net = build_volatile_network(update_type, &data);
    let explicit_net = build_explicit_network(update_type, &data);

    let label = format!("{} volatile vs explicit", update_type);

    // Input nodes should agree
    assert_value_level_match(&volatile_net, 0, &explicit_net, 0, &format!("{} input", label));

    // Value level of volatile node 1 should match explicit node 1
    assert_value_level_match(&volatile_net, 1, &explicit_net, 1, &label);

    // Volatility level of volatile node 1 should match explicit node 2
    assert_vol_level_match(&volatile_net, 1, &explicit_net, 2, &label);
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[test]
fn test_volatile_standard_matches_explicit() {
    compare_volatile_and_explicit("standard");
}

#[test]
fn test_volatile_ehgf_matches_explicit() {
    compare_volatile_and_explicit("eHGF");
}

#[test]
fn test_volatile_unbounded_matches_explicit() {
    compare_volatile_and_explicit("unbounded");
}
